homeassistant/components/github/coordinator.py

```diff
@@ -0,0 +1,142 @@
+"""Custom data update coordinators for the GitHub integration."""
+from __future__ import annotations
+
+import asyncio
+from typing import Literal, TypedDict
+
+from aiogithubapi import (
+    GitHubAPI,
+    GitHubCommitModel,
+    GitHubException,
+    GitHubIssueModel,
+    GitHubReleaseModel,
+    GitHubRepositoryModel,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, T
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DEFAULT_UPDATE_INTERVAL, DOMAIN, LOGGER, IssuesPulls
+
+CoordinatorKeyType = Literal["information", "release", "issue", "commit"]
+
+
+class GitHubBaseDataUpdateCoordinator(DataUpdateCoordinator[T]):
+    """Base class for GitHub data update coordinators."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        client: GitHubAPI,
+        repository: str,
+    ) -> None:
+        """Initialize GitHub data update coordinator base class."""
+        self.config_entry = entry
+        self.repository = repository
+        self._client = client
+
+        super().__init__(
+            hass,
+            LOGGER,
+            name=DOMAIN,
+            update_interval=DEFAULT_UPDATE_INTERVAL,
+        )
+
+    async def fetch_data(self) -> T:
+        """Fetch data from GitHub API."""
+
+    async def _async_update_data(self) -> T:
+        try:
+            return await self.fetch_data()
+        except GitHubException as exception:
+            LOGGER.exception(exception)
+            raise UpdateFailed(exception) from exception
+
+
+class RepositoryInformationDataUpdateCoordinator(
+    GitHubBaseDataUpdateCoordinator[GitHubRepositoryModel]
+):
+    """Data update coordinator for repository information."""
+
+    async def fetch_data(self) -> GitHubRepositoryModel:
+        """Get the latest data from GitHub."""
+        result = await self._client.repos.get(self.repository)
+        return result.data
+
+
+class RepositoryReleaseDataUpdateCoordinator(
+    GitHubBaseDataUpdateCoordinator[GitHubReleaseModel]
+):
+    """Data update coordinator for repository release."""
+
+    async def fetch_data(self) -> GitHubReleaseModel | None:
+        """Get the latest data from GitHub."""
+        result = await self._client.repos.releases.list(
+            self.repository, **{"params": {"per_page": 1}}
+        )
+        return result.data[0] if result.data else None
+
+
+class RepositoryIssueDataUpdateCoordinator(
+    GitHubBaseDataUpdateCoordinator[IssuesPulls]
+):
+    """Data update coordinator for repository issues."""
+
+    async def fetch_data(self) -> IssuesPulls:
+        """Get the latest data from GitHub."""
+
+        async def _get_issues():
+            response = await self._client.repos.issues.list(
+                self.repository, **{"params": {"per_page": 100}}
+            )
+            if not response.is_last_page:
```

Comment:
Do we really need to fetch and store that data? that's a lot of data for HA.

Reply:
Yes, we do.

Reply:
Found a way to limit calls and data handled/stored